### PR TITLE
Refactor(eos_designs): Add testcases for ipaddressing and move it to plugin utils

### PR DESCRIPTION
**Pull request by Vibhu-gslab**
## Change Summary

Writing unit testcase for get_ip_from_pool and move it to plugin_utils.

## Proposed changes

- Rename `_ip()` to `get_ip_from_pool()` and move to `plugin_utils.utils`
- Move more IP logic from `eos_designs` Python modules to `ip_addressing`

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly.
- [ ] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/plugins/plugin_utils/ip_addressing_utils.py
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/plugin_utils/ip_addressing_utils.py]: maybe call it `get_ip_from_pool` or something?

File: ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py
Comment (gmuloc) [inline ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: why do we import these?

File: ansible_collections/arista/avd/plugins/plugin_utils/ip_addressing_utils.py
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/plugins/plugin_utils/ip_addressing_utils.py]: sure, will change it.

File: ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: This is to support previous python versions as it is followed in other modules. Shall we remove it?

Comment (gmuloc) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: `print_function` as far as I know is to support `print(..)` format in python 2 rather than `print ...` so I am not sure we would really need it.

Let's maybe try to understand what these do nd see if we can get rid of them?

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: @gmuloc this is ansible boilerplate enforced by ansible-test for _everything_ that could potentially run on a target host. Depending on where we import this (in modules or filters) it will be enforced. @Vibhu-gslab you can try to remove it and see if it fails `ansible-test sanity`.

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py:1]: ah you did :)

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: same for metaclass

Comment (gmuloc) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: hmmm fair enough .. let's leave them in and wait for ansible to catch up on python2 deprecation.. apologies for the noise @Vibhu-gslab

Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: @gmuloc no worries :), I will still try to remove it and see if sanity passes as suggested by @ClausHolbechArista .
Can you please approve running workflows in the PR. Thanks.

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: what was the conclusion here?

Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: I am getting warning on running ansible-test sanity (WARNING: Skipping sanity test "compile" on Python 3.8 because it could not be found.) I have pushed the code after removing future and metaclass, to know for sure if we can run all CI checks it would be best (need approval for the same).

Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: All the tests have passed in CI after removing "future and metaclass" :)

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: perfect. Then we can clean up the other test files later. Thanks

File: ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py
Comment (gmuloc) [inline ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: so why do we parametrize if there is only one test case?

You can reformat this file with two functions:
one for valid test
one for invalid test with some parametrize for the four exceptions, using the expected exception string as another parameter.

To make the test have a meaningful name you can do:

```
@pytest.mark.parametrize("pool, prefixlen, subnet_offset, ip_offset, expected_raise", 
    [
        pytest.param("1.2.3.0/32", prefixlen, subnet_offset, ip_offset, <expected_raise>, id="test_ip_with_low_subnet_size")
    ]
)
def test_get_ip_from_pool_invalid(pool, prefixlen, subnet_offset, ip_offset, expected_raise):
    with pytest.raises(AristaAvdError) as exc_info:
        get_ip_from_pool(pool, prefixlen, subnet_offset, ip_offset)
    assert str(exc_info.value) == expected_raise
```

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/vxlan_interface.py
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/vxlan_interface.py]: so should we not have some kind of way to access get_ip_from_pool through the `self._avd_ip_addressing._ip` still? @ClausHolbechArista

File: ansible_collections/arista/avd/plugins/plugin_utils/ip_addressing_utils.py
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/ip_addressing_utils.py]: Can you please add a docstring?

Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/plugins/plugin_utils/ip_addressing_utils.py]: sure will do..

File: ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py
Comment (Vibhu-gslab) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: Updated the code.
Thanks

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/vxlan_interface.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/vxlan_interface.py]: @gmuloc I think the better solution here would be to create a method for `underlay_l3_multicast_group_ipv4` in `avd_ip_addressing`.

Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/vxlan_interface.py]: Agreed - @Vibhu-gslab could you please create a method for this in avd_ip_addressing and use it here? and same for l2multicast_group below?

File: ansible_collections/arista/avd/plugins/plugin_utils/utils/get_ip_from_pool.py
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/utils/get_ip_from_pool.py]: This does not look like our other docstrings

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/avdipaddressing.py
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/avdipaddressing.py:266]: This is not used anywhere for now

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/avdipaddressing.py
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/avdipaddressing.py:272]: so once this is implemented, is the `vrf_vni` param still needed? if not, should we then avoid passing both and let the caller decide which one to give?

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/vxlan_interface.py
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/vxlan_interface.py:98]: this is not really used today?

File: ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py
Comment (gmuloc) [inline ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: probably should be capitalized to be very pedantic :p

```suggestion
pool, prefixlen, subnet_offset, ip_offset = ("1.2.3.4/24", 32, 1, 0)
```

File: ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py
Comment (gmuloc) [inline ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: probably worth adding some valid border tests cases (like getting the last ip in a subnet either via offset and using some value of subnet_offset)

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/avdipaddressing.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/avdipaddressing.py:272]: In case someone use a custom ip addressing class this might be needed for their logic.

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/avdipaddressing.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/avdipaddressing.py:266]: I just wanted to keep the signature stable even when we change to id later.

File: ansible_collections/arista/avd/plugins/plugin_utils/utils/get_ip_from_pool.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/utils/get_ip_from_pool.py]: Fixed

File: ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/tests/unit/plugins/plugin_utils/test_ipaddressing.py]: Done

**Review by gmuloc — APPROVED** [review]
LGTM - once CI passes

**Review by carlbuchmann — APPROVED** [review]
LGTM

**Pull request merged**
